File: desktop-client/src/main/java/uni/cc4p1/client/model/MessageParser.java
package uni.cc4p1.client.model;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

public class MessageParser {

    public static Message readHeader(DataInputStream in) throws IOException {
        int payloadLen = in.readInt();       // 4 bytes
        byte typeCode = in.readByte();       // 1 byte
        short senderId = in.readShort();     // 2 bytes
        short receiverId = in.readShort();   // 2 bytes

        MessageType type = MessageType.fromCode(typeCode);
        return new Message(payloadLen, type, senderId, receiverId);
    }

    public static void writeMessage(DataOutputStream out, Message header, byte[] payload) throws IOException {
        out.writeInt(payload != null ? payload.length : 0);
        out.writeByte(header.type().getCode());
        out.writeShort(header.senderId());
        out.writeShort(header.receiverId());
        if (payload != null && payload.length > 0) {
            out.write(payload);
        }
        out.flush();
    }
}

File: mobile-client/app/src/main/java/uni/cc4p1/client/model/MessageParser.kt
package uni.cc4p1.client.model

import java.io.DataInputStream
import java.io.DataOutputStream

object MessageParser {

    fun readHeader(input: DataInputStream): Message {
        val payloadLen = input.readInt()
        val typeCode = input.readByte().toInt() and 0xFF
        val senderId = input.readShort()
        val receiverId = input.readShort()
        val type = MessageType.fromCode(typeCode)
        return Message(payloadLen, type, senderId, receiverId)
    }

    fun writeMessage(output: DataOutputStream, header: Message, payload: ByteArray?) {
        output.writeInt(payload?.size ?: 0)
        output.writeByte(header.type.code)
        output.writeShort(header.senderId.toInt())
        output.writeShort(header.receiverId.toInt())
        if (payload != null && payload.isNotEmpty()) {
            output.write(payload)
        }
        output.flush()
    }
}
